modules/schema/src/main/java/org/apache/ignite/internal/schema/BinaryRowUpgrader.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.schema;
+
+import org.apache.ignite.internal.schema.row.Row;
+import org.apache.ignite.internal.schema.row.RowAssembler;
+
+/** {@link BinaryRow} upgrader to the required schema version. */
+public class BinaryRowUpgrader {
+    private final SchemaRegistry schemaRegistry;
+
+    private final SchemaDescriptor targetSchema;
+
+    /** Constructor. */
+    public BinaryRowUpgrader(SchemaRegistry schemaRegistry, int targetSchemaVersion) {
+        this(schemaRegistry, schemaRegistry.schema(targetSchemaVersion));
+    }
+
+    /** Constructor. */
+    public BinaryRowUpgrader(SchemaRegistry schemaRegistry, SchemaDescriptor targetSchema) {
+        this.schemaRegistry = schemaRegistry;
+        this.targetSchema = targetSchema;
+    }
+
+    /**
+     * Returns an upgraded {@link BinaryRow} to the required schema version, or the {@code source} if its schema version is equal to the
+     * required schema version.
+     *
+     * @param source Source binary row.
+     */
+    public BinaryRow upgrade(BinaryRow source) {
+        if (source.schemaVersion() == targetSchema.version()) {
```

Comment:
It might happen that the source version is greater than the target version. For index building, this can only happen if a table was ALTERed between REGISTERED and BUILDING of the index and a transaction was started, tuple written and committed after the ALTER TABLE and before BUILDING. In such cases, we don't need to upgrade the row as it will actually be a downgrade. So we should skip upgrading for such cases.

I'm not sure if it's better to handle this inside this method or on the side of the caller.

Reply:
I think it’s better to handle this situation here, added a unit test.